sdk/core/azure-core/src/main/java/com/azure/core/implementation/ByteCountingWritableByteChannel.java

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation;
+
+import com.azure.core.util.ProgressReporter;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+/**
+ * Count bytes written to the target channel.
+ */
+public class ByteCountingWritableByteChannel implements WritableByteChannel {
```

Comment:
Right now this is only updating download, but for when upload is being updated would we just want this to be extending `ByteChannel` and have it function the same way (but synchronously) as `AsynchronousByteChannel`?

Reply:
I covered both for `AsynchronousByteChannel` variant because there's not much choice there, i.e. there's no `Writeable/ReadableAsynchronousByteChannel` . 

If we ever need to cover `ReadableByteChannel` we'd create `ByteCountingReadableByteChannel` and then `ByteCountingByteChannel` would be aggregate of both, i.e.

```
class ByteCountingByteChannel {
   private final ByteCountingReadableByteChannel delegateAsReadableChannel;
   private final ByteCountingWriteableByteChannel delegateAsWritableChannel;

  // delegate to the above delegates in impl respectively.
}
```

This would match granularity of what JDK has.